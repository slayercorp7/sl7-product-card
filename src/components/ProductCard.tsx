import styles from "../styles/styles.module.css";

import { useProduct } from "../hooks/useProduct";
import { createContext } from "react";
import {
  ProductContextProps,
  Product,
  onChangeArgs,
  InitialValues,
  ProductCardHandlers,
} from "../interfaces/interfaces";
import React from "react";

export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export interface Props {
  children: (args: ProductCardHandlers) => JSX.Element;
  product: Product;
  className?: String;
  style?: React.CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  InitialValues?: InitialValues;
}

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  InitialValues,
}: Props) => {
  const { counter, increaseBy, isMaxCountReached, reset } =
    useProduct({ onChange, product, value, InitialValues });

  return (
    <Provider value={{ counter, increaseBy, product }}>
      <div className={`${styles.productCard} ${className}`} style={style}>
        {children({
          count: counter,
          isMaxCountReached,
          maxCount: InitialValues?.maxCount,
          product,
          increaseBy,
          reset,
        })}
      </div>
    </Provider>
  );
};
