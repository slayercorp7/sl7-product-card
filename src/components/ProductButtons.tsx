import { useCallback, useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from "../styles/styles.module.css";
import React from 'react';

export interface Props {
  className?: string;
  style?: React.CSSProperties;
  maxCount?: number;
}

export const ProductButtons = ({className, style, maxCount} : Props) => {

  //maxcount, 
    const { increaseBy, counter } = useContext(ProductContext);

    // ismaxreached = usecallback,[counter, maxCount]
    const isMaxReached = useCallback(
      () => !!maxCount && counter === maxCount,
      [counter, maxCount],
    );
  
    //true count===maxcount
    //false si no 
    return (
      <div 
      className={`${styles.buttonsContainer} ${className}`}
      style={style}
      >
        <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
          -
        </button>
        <div className={styles.countLabel}> {counter} </div>
        <button className={`${styles.buttonAdd} ${isMaxReached() && styles.disabled}`} onClick={() => increaseBy(1)}>
          +
        </button>
      </div>
    );
  };