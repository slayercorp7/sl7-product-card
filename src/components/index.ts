
import { ProductCard as ProductCardHOC } from "./ProductCard";
import { ProductTitle } from "./ProductTitle";
import { ProductImage } from "./ProductImage";
import { ProductButtons } from "./ProductButtons";
import { PoductCardHOCProps } from "../interfaces/interfaces";
export { ProductButtons } from "./ProductButtons";
export { ProductImage } from "./ProductImage";
export { ProductTitle } from "./ProductTitle";

export const ProductCard: PoductCardHOCProps = Object.assign(ProductCardHOC, {
  Title: ProductTitle,
  Image: ProductImage,
  Buttons: ProductButtons,
});
