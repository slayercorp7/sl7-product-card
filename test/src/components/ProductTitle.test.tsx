import * as React from 'react';
import renderer from 'react-test-renderer';
import { ProductTitle } from '../../../src/components/ProductTitle';
import { ProductCard } from '../../../src/components/ProductCard';
import { product1 } from '../../data/products';

describe('pruebas en Product Title', () => {
  test('debe de mostrar el componente correctamente con el titulo personalizado', () => {
    const wrapper = renderer.create(<ProductTitle title="hola" />);
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
  test('debe de mostrar el componente con el nombre del producto', () => {
    const wrapper = renderer.create(
      <ProductCard product={product1}>{() => <ProductTitle />}</ProductCard>
    );
    expect(wrapper.toJSON()).toMatchSnapshot();
  });
});
