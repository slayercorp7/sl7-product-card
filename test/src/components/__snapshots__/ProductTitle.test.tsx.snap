// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en Product Title debe de mostrar el componente con el nombre del producto 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined"
  >
    compra
  </span>
</div>
`;

exports[`pruebas en Product Title debe de mostrar el componente correctamente con el titulo personalizado 1`] = `
<span
  className="productDescription undefined"
>
  hola
</span>
`;
